docs/contribute/bmi.md

```diff
@@ -0,0 +1,44 @@
+`finalize()`              | Write all results to the configured files
+`get_current_time()`      | Get the current time of the Ribasim simulation
+`get_end_time()`          | Get the final time of the Ribasim simulation in seconds
+`get_start_time()`        | Get the start time of the Ribasim simulation (`0.0`)
+`get_time_step()`         | Get the proposed next internal Ribasim timestep
+`get_time_units()`        | Get the time unit (`s`)
+`get_value_ptr(string)`   | Get the pointer to a Ribasim internal array (see below)
+`update()`                | Set a single Ribasim internal time step
+`update_until(time)`      | Set Ribasim internal timesteps until the specified time
+
+Depending on what is specified in the Ribasim TOML configuration file, Ribasim can internally have adaptive (non-constant) timesteps. `update_until` will always try to progress the Ribasim simulation to exactly the time specified. This however can fail for algorithms that only support a fixed timestep if that timestep does not fit into the interval until the specified time an integer amount of times.
+
+## Data pointers
+
+The following internal array data is exposed via the BMI using `get_value_ptr(string)`:
+
+string                          | meaning                                | type    | unit         | sorted by
+------------------------------- | -------------------------------------- | ------- | ------------ | ----------
+`basin.storage`                 | storage per basin                      | Float64 | $m^3$        | basin node ID
+`basin.level`                   | level per basin                        | Float64 | $m$          | basin node ID
+`basin.infiltration`            | infiltration per basin                 | Float64 | $m^3 s^{-1}$ | basin node ID
+`basin.drainage`                | drainage per basin                     | Float64 | $m^3 s^{-1}$ | basin node ID
+`basin.infiltration_integrated` | time integrated infiltration per basin | Float64 | $m^3$        | basin node ID
+`basin.drainage_integrated`     | time integrated drainage per basin     | Float64 | $m^3$        | basin node ID
+`basin.subgrid_level`           | subgrid level                          | Float64 | $m$          | subgrid ID
+`user_demand.demand`            | demand per node ID per priority        | Float64 | $m^3 s^{-1}$ | user_demand node ID, priority index
+`user_demand.realized`          | time integrated intake flow per user   | Float64 | $m^3$        | user_demand node ID
+
+Additional notes:
+
+- `user_demand.demand` yields the only 2D array, the other arrays are 1D. This array is indexed as `(node_idx, priority_idx)`
+- The index of e.g. basins and user demand nodes needs to be inferred from the Ribasim input. The same holds for `priority_idx`, which is global over all subnetworks
+- The `*_integrated` values are integrated from the start of the simulation onward. It might be beneficial to reset this to $0$ at certain times for long simulations to avoid loss of accuracy.
```

Comment:
I would recommend to add a note indicating that basin.infiltration and basin.drainage are averaged over the timestep (if I am correct from t-1 tot t)

Reply:
That is different between what is exposed via the BMI and what is in the output. The output gives averaged values, the BMI exposes an integrated value

Reply:
That way the coupler has the freedom to decide over which period to average

Reply:
you already indicate that `*_integrated' is the accumulation.
but your description of the other variables can be more precise, by indicating it is instantaneous, averaged or whatever it is. It now is still a bit open-ended